File: sdk/typescript/src/namespaces/belief-network.ts
/**
 * Belief Network Namespace API
 *
 * Provides access to debate belief networks, cruxes, and load-bearing claims.
 * These represent the epistemic structure of arguments within debates.
 */

/**
 * A crux - a key disagreement point that, if resolved, would change positions
 */
export interface Crux {
  id: string;
  debate_id: string;
  description: string;
  agents_pro: string[];
  agents_con: string[];
  importance: number;
  resolved: boolean;
  resolution?: string;
  identified_at: string;
}

/**
 * A load-bearing claim - a claim that many other claims depend on
 */
export interface LoadBearingClaim {
  id: string;
  debate_id: string;
  claim: string;
  agent_name: string;
  confidence: number;
  dependents_count: number;
  dependent_claims: string[];
  supporting_evidence: string[];
  created_at: string;
}

/**
 * Belief network node
 */
export interface BeliefNode {
  id: string;
  claim: string;
  agent_name: string;
  confidence: number;
  type: 'premise' | 'inference' | 'conclusion';
  supporting_nodes: string[];
  opposing_nodes: string[];
}

/**
 * Belief network edge
 */
export interface BeliefEdge {
  source_id: string;
  target_id: string;
  relationship: 'supports' | 'contradicts' | 'qualifies' | 'depends_on';
  strength: number;
}

/**
 * Belief network graph node for visualization
 */
export interface BeliefGraphNode {
  id: string;
  claim_id: string;
  statement: string;
  author: string;
  centrality: number;
  is_crux: boolean;
  crux_score: number | null;
  entropy: number;
  belief: {
    true_prob: number;
    false_prob: number;
    uncertain_prob: number;
  } | null;
}

/**
 * Belief network graph edge for visualization
 */
export interface BeliefGraphEdge {
  source: string;
  target: string;
  weight: number;
  type: 'supports' | 'contradicts' | 'qualifies' | 'depends_on' | 'influences';
}

/**
 * Belief network graph response
 */
export interface BeliefGraphResponse {
  nodes: BeliefGraphNode[];
  links: BeliefGraphEdge[];
  metadata: {
    debate_id: string;
    total_claims: number;
    crux_count: number;
  };
}

/**
 * Export format types
 */
export type ExportFormat = 'json' | 'graphml' | 'csv';

/**
 * JSON export response
 */
export interface JsonExportResponse {
  format: 'json';
  debate_id: string;
  nodes: Array<{
    id: string;
    statement: string;
    author: string;
    centrality: number;
    is_crux: boolean;
  }>;
  edges: Array<{
    source: string;
    target: string;
    weight: number;
    type: string;
  }>;
  summary: {
    total_nodes: number;
    total_edges: number;
    crux_count: number;
  };
}

/**
 * GraphML export response
 */
export interface GraphMLExportResponse {
  format: 'graphml';
  debate_id: string;
  content: string;
  content_type: 'application/xml';
}

/**
 * CSV export response
 */
export interface CSVExportResponse {
  format: 'csv';
  debate_id: string;
  nodes_csv: Array<Record<string, unknown>>;
  edges_csv: Array<Record<string, unknown>>;
  headers: {
    nodes: string[];
    edges: string[];
  };
}

/**
 * Union type for export responses
 */
export type ExportResponse = JsonExportResponse | GraphMLExportResponse | CSVExportResponse;

/**
 * Interface for the internal client used by BeliefNetworkAPI.
 */
interface BeliefNetworkClientInterface {
  get<T>(path: string): Promise<T>;
  request<T>(method: string, path: string, options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }): Promise<T>;
}

/**
 * Belief Network API namespace.
 *
 * Provides methods for analyzing debate belief structures:
 * - Identify cruxes (key disagreements)
 * - Find load-bearing claims
 * - Explore argument dependencies
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Get cruxes for a debate
 * const { cruxes } = await client.beliefNetwork.getCruxes('debate-123');
 *
 * // Get load-bearing claims
 * const { claims } = await client.beliefNetwork.getLoadBearingClaims('debate-123');
 * ```
 */
export class BeliefNetworkAPI {
  constructor(private client: BeliefNetworkClientInterface) {}

  /**
   * Get cruxes (key disagreements) for a debate.
   *
   * Cruxes are points where, if one party changed their mind,
   * it would significantly affect their overall position.
   */
  async getCruxes(debateId: string, options?: { resolved?: boolean; min_importance?: number }): Promise<{ cruxes: Crux[]; total: number }> {
    return this.client.request('GET', `/api/v1/belief-network/${debateId}/cruxes`, { params: options });
  }

  /**
   * Get load-bearing claims for a debate.
   *
   * Load-bearing claims are claims that many other claims depend on.
   * If these are invalidated, the overall argument structure may collapse.
   */
  async getLoadBearingClaims(debateId: string, options?: { min_dependents?: number; agent?: string }): Promise<{ claims: LoadBearingClaim[]; total: number }> {
    return this.client.request('GET', `/api/v1/belief-network/${debateId}/load-bearing-claims`, { params: options });
  }

  /**
   * Get belief network as a graph structure for visualization.
   *
   * Returns nodes (claims) and links (influence relationships) suitable
   * for force-directed graph rendering.
   *
   * @param debateId - The debate ID
   * @param options.include_cruxes - Include crux analysis in nodes (default: true)
   *
   * @example
   * ```typescript
   * const graph = await client.beliefNetwork.getGraph('debate-123');
   * console.log(`${graph.nodes.length} claims, ${graph.links.length} relationships`);
   * console.log(`${graph.metadata.crux_count} cruxes identified`);
   * ```
   */
  async getGraph(debateId: string, options?: { include_cruxes?: boolean }): Promise<BeliefGraphResponse> {
    return this.client.request('GET', `/api/v1/belief-network/${debateId}/graph`, { params: options });
  }

  /**
   * Export belief network in various formats.
   *
   * Supported formats:
   * - json: Full JSON structure with nodes, edges, and summary
   * - graphml: GraphML format for Gephi/yEd visualization tools
   * - csv: CSV-friendly structure with separate nodes and edges arrays
   *
   * @param debateId - The debate ID
   * @param format - Export format ('json' | 'graphml' | 'csv')
   *
   * @example
   * ```typescript
   * // Export as GraphML for Gephi
   * const graphml = await client.beliefNetwork.export('debate-123', 'graphml');
   * console.log(graphml.content); // XML content
   *
   * // Export as JSON
   * const json = await client.beliefNetwork.export('debate-123', 'json');
   * console.log(`${json.summary.total_nodes} nodes`);
   * ```
   */
  async export(debateId: string, format: ExportFormat = 'json'): Promise<ExportResponse> {
    return this.client.request('GET', `/api/v1/belief-network/${debateId}/export`, { params: { format } });
  }
}

File: sdk/python/aragora/namespaces/belief_network.py
"""
Belief Network Namespace API.

Provides access to debate belief networks, cruxes, and load-bearing claims.
These represent the epistemic structure of arguments within debates.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypedDict

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

ExportFormat = Literal["json", "graphml", "csv"]
NodeType = Literal["premise", "inference", "conclusion"]
RelationshipType = Literal["supports", "contradicts", "qualifies", "depends_on", "influences"]


class Crux(TypedDict, total=False):
    """A crux - a key disagreement point that would change positions if resolved."""

    id: str
    debate_id: str
    description: str
    agents_pro: list[str]
    agents_con: list[str]
    importance: float
    resolved: bool
    resolution: str | None
    identified_at: str


class LoadBearingClaim(TypedDict, total=False):
    """A load-bearing claim - a claim that many other claims depend on."""

    id: str
    debate_id: str
    claim: str
    agent_name: str
    confidence: float
    dependents_count: int
    dependent_claims: list[str]
    supporting_evidence: list[str]
    created_at: str


class BeliefNode(TypedDict, total=False):
    """Belief network node."""

    id: str
    claim: str
    agent_name: str
    confidence: float
    type: str
    supporting_nodes: list[str]
    opposing_nodes: list[str]


class BeliefGraphNode(TypedDict, total=False):
    """Belief network graph node for visualization."""

    id: str
    claim_id: str
    statement: str
    author: str
    centrality: float
    is_crux: bool
    crux_score: float | None
    entropy: float
    belief: dict[str, float] | None


class BeliefGraphEdge(TypedDict, total=False):
    """Belief network graph edge for visualization."""

    source: str
    target: str
    weight: float
    type: str


class BeliefGraphResponse(TypedDict, total=False):
    """Belief network graph response."""

    nodes: list[BeliefGraphNode]
    links: list[BeliefGraphEdge]
    metadata: dict[str, Any]


class BeliefNetworkAPI:
    """
    Synchronous Belief Network API.

    Provides methods for analyzing debate belief structures:
    - Identify cruxes (key disagreements)
    - Find load-bearing claims
    - Explore argument dependencies
    - Export network visualizations

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> # Get cruxes for a debate
        >>> result = client.belief_network.get_cruxes("debate-123")
        >>> for crux in result["cruxes"]:
        ...     print(f"Crux: {crux['description']} (importance: {crux['importance']})")
        >>> # Get load-bearing claims
        >>> claims = client.belief_network.get_load_bearing_claims("debate-123")
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    def get_cruxes(
        self,
        debate_id: str,
        resolved: bool | None = None,
        min_importance: float | None = None,
    ) -> dict[str, Any]:
        """
        Get cruxes (key disagreements) for a debate.

        Cruxes are points where, if one party changed their mind,
        it would significantly affect their overall position.

        Args:
            debate_id: The debate ID.
            resolved: Filter by resolution status.
            min_importance: Minimum importance score (0.0-1.0).

        Returns:
            Dict with 'cruxes' list and 'total' count.
        """
        params: dict[str, Any] = {}
        if resolved is not None:
            params["resolved"] = resolved
        if min_importance is not None:
            params["min_importance"] = min_importance
        return self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/cruxes",
            params=params if params else None,
        )

    def get_load_bearing_claims(
        self,
        debate_id: str,
        min_dependents: int | None = None,
        agent: str | None = None,
    ) -> dict[str, Any]:
        """
        Get load-bearing claims for a debate.

        Load-bearing claims are claims that many other claims depend on.
        If these are invalidated, the overall argument structure may collapse.

        Args:
            debate_id: The debate ID.
            min_dependents: Minimum number of dependent claims.
            agent: Filter by agent name.

        Returns:
            Dict with 'claims' list and 'total' count.
        """
        params: dict[str, Any] = {}
        if min_dependents is not None:
            params["min_dependents"] = min_dependents
        if agent:
            params["agent"] = agent
        return self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/load-bearing-claims",
            params=params if params else None,
        )

    def get_graph(
        self,
        debate_id: str,
        include_cruxes: bool = True,
    ) -> BeliefGraphResponse:
        """
        Get belief network as a graph structure for visualization.

        Returns nodes (claims) and links (influence relationships) suitable
        for force-directed graph rendering.

        Args:
            debate_id: The debate ID.
            include_cruxes: Include crux analysis in nodes (default: True).

        Returns:
            BeliefGraphResponse with nodes, links, and metadata.
        """
        params: dict[str, Any] = {}
        if not include_cruxes:
            params["include_cruxes"] = False
        return self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/graph",
            params=params if params else None,
        )

    def export(
        self,
        debate_id: str,
        format: ExportFormat = "json",
    ) -> dict[str, Any]:
        """
        Export belief network in various formats.

        Supported formats:
        - json: Full JSON structure with nodes, edges, and summary
        - graphml: GraphML format for Gephi/yEd visualization tools
        - csv: CSV-friendly structure with separate nodes and edges arrays

        Args:
            debate_id: The debate ID.
            format: Export format ('json', 'graphml', or 'csv').

        Returns:
            Export response with format-specific content.
        """
        return self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/export",
            params={"format": format},
        )


class AsyncBeliefNetworkAPI:
    """Asynchronous Belief Network API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    async def get_cruxes(
        self,
        debate_id: str,
        resolved: bool | None = None,
        min_importance: float | None = None,
    ) -> dict[str, Any]:
        """Get cruxes (key disagreements) for a debate."""
        params: dict[str, Any] = {}
        if resolved is not None:
            params["resolved"] = resolved
        if min_importance is not None:
            params["min_importance"] = min_importance
        return await self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/cruxes",
            params=params if params else None,
        )

    async def get_load_bearing_claims(
        self,
        debate_id: str,
        min_dependents: int | None = None,
        agent: str | None = None,
    ) -> dict[str, Any]:
        """Get load-bearing claims for a debate."""
        params: dict[str, Any] = {}
        if min_dependents is not None:
            params["min_dependents"] = min_dependents
        if agent:
            params["agent"] = agent
        return await self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/load-bearing-claims",
            params=params if params else None,
        )

    async def get_graph(
        self,
        debate_id: str,
        include_cruxes: bool = True,
    ) -> BeliefGraphResponse:
        """Get belief network as a graph structure for visualization."""
        params: dict[str, Any] = {}
        if not include_cruxes:
            params["include_cruxes"] = False
        return await self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/graph",
            params=params if params else None,
        )

    async def export(
        self,
        debate_id: str,
        format: ExportFormat = "json",
    ) -> dict[str, Any]:
        """Export belief network in various formats."""
        return await self._client.request(
            "GET",
            f"/api/v1/belief-network/{debate_id}/export",
            params={"format": format},
        )
